Potential bug fix in kepclip and added TPF support.

christinahedges opened the pull request:
kepclip's 'reject' argument seemed to be the wrong way round to me, I put it back. Added support for clipping TPFs.

mirca commented:
Thank you @christinahedges! Let's make those failing tests pass =D

mirca commented on pyke/kepclip.py:
```diff
@@ -1,5 +1,5 @@
-from .utils import PyKEArgumentHelpFormatter
-from . import kepio, kepmsg, kepkey
+from pyke.utils import PyKEArgumentHelpFormatter
+from pyke import kepio, kepmsg, kepkey
```
no need to change these lines

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	h=fits.open('data/dip_llc.fits')
+	t=h[1].data['TIME'][200:400]
+	ranges='{}'.format(t[0]+2454833)+',{}'.format(t[-1]+2454833)
+	print ranges
```
use ``()`` around ``ranges`` :)

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	kepclip(infile='data/dip_llc.fits',
+	    outfile='data/out.fits',
+	    ranges=ranges,
+	    overwrite=True)
```
This could fit in one line, I guess

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	    outfile='data/out.fits',
+	    ranges=ranges,
+	    overwrite=True)
+	j=fits.open('data/out.fits')
```
nitpick, but useful, add whitespaces before and after math symbols

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	'''
+	Test the tpf clipping
+	Uses a clipped TPF of Tabby's star.
+	'''
```
this can be shortened in one line "Test the tpf clipping. Uses a clipped TPF of Tabby's star."

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	h=fits.open('data/testtpf.fits')
+	t=h[1].data['TIME'][0:2]
+	ranges='{0:.15f}'.format(t[0]+2454833)+',{0:.15f}'.format(t[-1]+2454833)
+	
```
same nitpick here: white spaces between math symbols and after commas

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -0,0 +1,70 @@
+	kepclip(infile='data/testtpf.fits',
+		outfile='data/out.fits',
+		ranges=ranges,
+		overwrite=True)
```
fit these line in one?

mirca commented on pyke/kepclip.py:
```diff
@@ -98,111 +98,149 @@ def kepclip(infile, ranges, outfile=None, datacol='SAP_FLUX', plot=False,
+            if ymin - yr * 0.01 <= 0.0:
+                plt.ylim(1.0e-10, ymax + yr * 0.01)
+
+    if filetype=='TPF':
```
white spaces between ``==``

mirca commented on pyke/kepclip.py:
```diff
@@ -98,111 +98,149 @@ def kepclip(infile, ranges, outfile=None, datacol='SAP_FLUX', plot=False,
+
+        # write output file
+
+        instr[1].data = table[:naxis2]
```
can you delete these blank lines above?

mirca commented on pyke/kepclip.py:
```diff
@@ -98,111 +98,149 @@ def kepclip(infile, ranges, outfile=None, datacol='SAP_FLUX', plot=False,
 
+
+
+
```
no need to have blank lines here

christinahedges commented:
Thanks @mirca that was a neat catch. Filled the test with spaces for you :)

christinahedges commented:
I've changed kepclip to remove the NaN removal function. I spoke with @mirca about this and my reasoning is that it's not the *expected* behavior of kepclip, so it shouldn't do it. Instead, there should be a routine that is dedicated to removing NaN values from fits files: kepclean.

I've added this routine. It takes in LCs and TPFs and passes back a new fits file that is identical, except for having cadences that were entirely NaN values removed. There is a new unit test for kepclean.

mirca commented on pyke/kepclean.py:
```diff
@@ -0,0 +1,161 @@
+            ydat = instr[ext].data[y]
+            if len(np.shape(ydat)) > 1:
+                for i in range(len(np.shape(ydat))-1):
+                    bad = np.all(np.isfinite(ydat)==False,axis=-1)
```
``np.isfinite(ydat)==False`` this is not recommend. Instead do, ``np.isfinite(ydat) is False``.

mirca replied:
I think you could even do ``bad = np.all(~np.isfinite(ydat),axis=-1)``?

mirca replied:
Reference: https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

mirca commented on pyke/kepclip.py:
```diff
@@ -90,108 +90,155 @@ def kepclip(infile, ranges, outfile=None, datacol='SAP_FLUX', plot=False,
     # read time and flux columns
     barytime = kepio.readtimecol(infile, table, logfile, verbose)
-    flux = kepio.readfitscol(infile, table, datacol, logfile, verbose)
+
```
could you remove those blank lines?

christinahedges replied:
blank lines








forever.

mirca commented on pyke/kepclean.py:
```diff
@@ -0,0 +1,161 @@
+                    bad = np.all(np.isfinite(ydat)==False,axis=-1)
+                    ydat = np.nansum(ydat, axis=-1)
+                    ydat[bad] = np.nan
+            bad = set(list(np.where(np.isfinite(ydat)==False)[0]))
```
Same thing here

mirca commented on pyke/kepclean.py:
```diff
@@ -0,0 +1,161 @@
+        errmsg = 'ERROR -- KEPCLIP: ' + outfile + ' exists. Use --overwrite'
+        kepmsg.err(logfile, errmsg, verbose)
+
+
```
Blank lines XD

mirca commented on pyke/kepclean.py:
```diff
@@ -0,0 +1,161 @@
+        if zero:
+            for d in instr[ext].data.names:
+                y = instr[ext].data[d]
+                y[np.isfinite(y) == False] = 0
```
Same thing here ``y[np.isfinite(y) == False] = 0`` ---> ``y[~np.isfinite(y)] = 0``

mirca commented on pyke/tests/test_kepclip.py:
```diff
@@ -1,22 +1,68 @@
+	assert len(h) == len(j)
+
+def test_kepclip_tpf():
+
```
blank line XD

mirca commented:
This is looking great! Thanks for revitalizing this PR and adding a new feature! 🍓 🍓 🕷 🍓 🕷

mirca commented on pyke/kepclean.py:
```diff
@@ -88,7 +88,7 @@ def kepclean(infile, outfile=None, zero=True, overwrite=False, verbose=False,
                     continue
         flux_keys = np.asarray(flux_keys,dtype=bool)
-        if np.any([(len(flux_keys) == 0),(len(np.where(flux_keys==True)[0]) == 0)]) :
+        if np.any([(len(flux_keys) == 0),(len(np.where(flux_keys == True)[0]) == 0)]) :
```
Add a blank space after the comma here. No need to a add a blank space before ``:``.
Also, change ``np.where(flux_keys == True)`` to ``np.where(flux_keys)``.

mirca commented on pyke/kepclean.py:
```diff
@@ -88,7 +88,7 @@ def kepclean(infile, outfile=None, zero=True, overwrite=False, verbose=False,
                 except:
                     continue
         flux_keys = np.asarray(flux_keys,dtype=bool)
```
can you add a blank space after the comma here?

mirca commented on pyke/kepclip.py:
```diff
@@ -98,12 +98,12 @@ def kepclip(infile, ranges, outfile=None, datacol='SAP_FLUX', plot=False,
         filetype='LC'
-    if len(set(instr[1].data.columns.names) & set(['SAP_FLUX']))==0:
+    if len(set(instr[1].data.columns.names) & set(['SAP_FLUX'])) == 0:
         filetype='TPF'
```
add white spaces betwen ``=``

christinahedges commented:
⚡️ 💀

christinahedges commented:
@barentsen Can you verify you're happy with this? kepclean is new and we don't want to sign off without you.

barentsen commented:
Catching up with e-mail -- I have two questions about `kepclean`:
1) Do you think the tool is necessary? Are there pyke tools which require NaN's to be removed in advance?  And if so, is it possible to make those tools robust against the presence of NaN's, instead of making the user take the extra step of running `kepclean`?  (We don't want to add and maintain a tool that nobody will ever use.)
2) If `kepclean` only removes NaN's we may want to think about a different name (`kepvacuum`?), because `kepclean` to me sounds like a tool that also removes outliers and cadences with bad quality flags?  Or is that perhaps what we envisage `kepclean` to become?

christinahedges commented:
> Do you think the tool is necessary? 

Yeah I do:
* For people who are running these routines in stages I think it's worth having each routine do exactly what it says, rather than having some tools secretly remove NaN's for you (flagging it in the header).
* This has better functionality. This also removes all the NaN values from every extension with FLUX, not just from one user specified extention as in some of the 'hidden' nan finding routines. Also TPFs have cadences where all values are NaN removed.
* Visualizations and plots that will be clearer by not having NaN values or zero'd values, although Kepdraw is robust against NaNs. 

I think this is a valuable tool, even if we just use it inside other keproutines.

>If kepclean only removes NaN's we may want to think about a different name (kepvacuum?)

It can absolutely be called kepvacuum, I like that even more.

christinahedges commented:
Geert has pointed out that kepclean would be more useful to users if there were functions to remove bad outliers and bad thruster firings.

barentsen commented:
Ok.  I think we can merge this if all of @mirca's comments were addressed, and open a separate issue to fully define the purpose of `kepclean` as a command-line tool.

christinahedges commented:
So close to a t-shirt! :shirt: @mirca you happy with that?
